src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -60,58 +65,340 @@ public class ResponseCachingMiddleware
+                return true;
+            }
+
+            return false;
+        }
+
+        internal static bool ConditionalRequestSatisfied(ResponseCachingContext context)
+        {
+            var cachedResponseHeaders = context.CachedResponseHeaders;
+            var ifNoneMatchHeader = context.TypedRequestHeaders.IfNoneMatch;
+
+            if (ifNoneMatchHeader != null)
+            {
+                if (ifNoneMatchHeader.Count == 1 && ifNoneMatchHeader[0].Equals(EntityTagHeaderValue.Any))
+                {
+                    return true;
+                }
+
+                if (cachedResponseHeaders.ETag != null)
+                {
+                    foreach (var tag in ifNoneMatchHeader)
                     {
-                        cachingContext.UnshimResponseStream();
+                        if (cachedResponseHeaders.ETag.Compare(tag, useStrongComparison: true))
+                        {
+                            return true;
+                        }
                     }
                 }
+            }
+            else if ((cachedResponseHeaders.LastModified ?? cachedResponseHeaders.Date) <= context.TypedRequestHeaders.IfUnmodifiedSince)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        internal static void FinalizeCachingHeaders(ResponseCachingContext context)
+        {
+            if (context.CacheabilityValidator.IsResponseCacheable(context))
+            {
+                context.ShouldCacheResponse = true;
+                context.StorageBaseKey = context.KeyProvider.CreateStorageBaseKey(context);
+
+                // Create the cache entry now
+                var response = context.HttpContext.Response;
+                var varyHeaderValue = response.Headers[HeaderNames.Vary];
+                var varyParamsValue = context.HttpContext.GetResponseCachingFeature()?.VaryParams ?? StringValues.Empty;
+                context.CachedResponseValidFor = context.ResponseCacheControlHeaderValue.SharedMaxAge
+                    ?? context.ResponseCacheControlHeaderValue.MaxAge
+                    ?? (context.TypedResponseHeaders.Expires - context.ResponseTime)
+                    ?? TimeSpan.FromSeconds(10);
+
+                // Check if any vary rules exist
+                if (!StringValues.IsNullOrEmpty(varyHeaderValue) || !StringValues.IsNullOrEmpty(varyParamsValue))
+                {
+                    // Normalize order and casing of vary by rules
+                    var normalizedVaryHeaderValue = GetNormalizedStringValues(varyHeaderValue);
+                    var normalizedVaryParamsValue = GetNormalizedStringValues(varyParamsValue);
+
+                    // Update vary rules if they are different
+                    if (context.CachedVaryRules == null ||
+                        !StringValues.Equals(context.CachedVaryRules.Params, normalizedVaryParamsValue) ||
```

Comment:
We need to normalize order though, `StringValues` contains an underlying array. Do you recommend sorting the underlying array when a certain `StringComparer` is set? That modifies the original value so I'm not sure if that's a good idea.